sdk/python/kfp/dsl/_container_op.py

```diff
@@ -869,7 +869,26 @@ def set_retry(self, num_retries: int):
           num_retries: Number of times to retry on failures.
         """
 
-        self.num_retries = num_retries
+        self.set_retry_strategy(num_retries=num_retries)
+        return self
+
+    def set_retry_strategy(
+        self,
+        num_retries: int = 0,
+        retry_on_error: bool = False,
```

Comment:
I think it worths some documentation on the behavior here. I am assuming if not retry_on_error and not retry_on_failure then it's Always, right? What if we both of them are true?

Also perhaps we can use a 'Enum' string here to make it more explicit, instead of relying on branching the logic? WDYT

Reply:
Reading from above it looks like to be `Both`, but I am wondering if it will be better to directly use Argo enum here, so that we don't overcomplicate things.

Reply:
>if not retry_on_error and not retry_on_failure then it's Always, right

If both are false, then there is no retry...
If both are true, then it will be "Both" in Argo.

>perhaps we can use a 'Enum' string here

I'm a bit concerned about mirroring Argo too closely and also about Enum usability in Python (you need to pass a particular imported object instead of just a boolean).

Reply:
> If both are false, then there is no retry...

Sorry you're exactly right :)

> I'm a bit concerned about mirroring Argo too closely and also about Enum usability in Python.

Let me explain a little bit. I am actually thinking making a `frozenset` of `Both, Always, ...`, then let user pass in a string to this function. Inside the function we can do some validation by comparing what's actually passed in vs. what was pre-defined.

WDYT? Also, can you elaborate on what is the concern about mirroring Argo here? Thanks

Reply:
+1 on having the user set `retry_policy` directly with a string value (validated by the compiler using a `frozenset`) so it can also receive a pipeline parameter

Reply:
>let user pass in a string to this function
>WDYT? Also, can you elaborate on what is the concern about mirroring Argo here?

We're exposing Argo constants directly to the component/pipeline authors without any ability to control, proxy or translate the options. Tekton probably names the options differently. If we go with magic strings, the SDK would be less portable.
I also prefer decomposing options into orthogonal parameters when possible. In this case the option perfectly decomposes into two booleans. If there were many options without any logical decomposition, I'd consider enums or strings.

> user set retry_policy directly with a string value (validated by the compiler using a frozenset) so it can also receive a pipeline parameter

I think these two parts are mutually exclusive. The compiler cannot validate a runtime value.